rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,482 @@
+will also expose a gRPC method to the client so the client can request the
+manifest file. The client will compare the manifest file to the introspection
+data to test that the nodes were created as expected.
+
+The next step would be to extend the planner node to support per-label dynamic
+node creation. This will be used to implement an alternative version of the
+private database example as a proof of concept for per-user node creation and
+gRPC invocation routing.
+
+Once the intial implementations are done, some of the common reusable code will
+be moved to the SDK. The fist example of such reusable code is support for node
+to accept an initial proto that contains a list of `Sender`s and `Receiver`s
+that wrap the handles for channels that should be available to the node, and
+potentially additional configuration files.
+
+The next step would be to update the existing examples to reuse the
+configuration-based `planner` node.
+
+## Future Work
+
+### Static analysis tools
+
+We can support application developers in finding potential issues in their
+applications through static analysis tools that can help detect common problems:
+
+- Check for consistency between application configuration and node graph
+  manifest
+- Check that all node and channel names are unique
+- Check that there is only one gRPC invocation receiver for each label
+- Check that there is only one HTTP invocation receiver for each label
+- Check that all nodes can write to handles they receive
+- Check that all nodes can read from handles they receive
+- Check that the node graph manifest does not violate defense-in-depth
+  restritcions
+- Hihglight points where data is declassified to `Public`
+- Find nodes that can never receive data (unneeded nodes)
+- Find nodes where data will get stuck (dead ends)
+- Find risky patterns (e.g. diamond shaped graph with declassification in each
+  path)
+
+Note: The static analysis tools that focus on the structure of the node graph
+will only be able to analyse the initial state. It is possible for nodes to send
+channel handles to other nodes embedded in protobuf messages, which would mean
+that additional connections might be created over time. Nodes could also close
+channels and nodes could terminate. This means that while the initial setup
+might be static it does not guarantee that the node structure will never change.
+
+### Declaration of supported message types
+
+Each node will be able to declare which protobuf message types it can read and
+write. Each channel will be annotated with the single message type it is
+intended for. If a channel should carry multiple message types, it can be done
+using a wrapper message and `oneof`. This will allow us to build additional
+static analysis tools to make sure each node can handle the messages it will
+receive and all messages that it might send will go to a node that can
+understand those messages.
```

Comment:
Note that there are two (mostly independent) static check layers: one is a regular type check, to make sure the types of the messages over channels and handled by nodes line up; the other is the "label check", which makes sure communication follows the structure of the IFC lattice. Perhaps clarify this here, and show some errors of things that may go wrong in either case, and how they may be caught (does not need to have the full logic of the checkers, but just to show what they would check).